stab_controller_multi_lqr_node.cpp

```diff
@@ -0,0 +1,416 @@
+    input_delays_left << input_delays_left.tail(NO_OF_DELAYS-1), input_left(0);
+    input_delays_right << input_delays_right.tail(NO_OF_DELAYS-1), input_right(0);
+
+
+    //Debugging outputs
+    //std::cout<<"input_left:"<<input_left<<std::endl;
+    //std::cout<<"input_right:"<<input_right<<std::endl;
+    //controller.PrintK();
+    //std::cout << "states_left" << state_left << '\n';
+    //std::cout << "states_right" << state_right << '\n';
+
+    // Publish maxon actutation inputs for debugging frequency
+    ascento_msgs::wheel_input_torque output_msg;
+    output_msg.torque_left = -input_left(0);
+    output_msg.torque_right = input_left(0);
+    maxon_execution_pub.publish(output_msg);
+  }
+
+  // Matrix Interpolation
+  void StabControllerMultiLqrNode::LipKMatrix(const double phi){
+    // method interpolates linearly betwen nearest K-matrices
+    int index1;
+    int index2;
+    double phi_normalised;
+
+    if (phi<phi_min) {
+      index1 = 0;
+      index2 = 0;
+      ROS_WARN("LIP: phi out of bounds, using phi min");
+
+    } else if (phi>phi_max) {
+      index1 = NO_OF_EQUILIBRIUM_POSITIONS-1;
+      index2 = NO_OF_EQUILIBRIUM_POSITIONS-1;
+      ROS_WARN("LIP: phi out of bounds, using phi max");
+
+    } else {
+      // map theta to [0, 1]:
+      phi_normalised = (phi-phi_min)/(phi_max-phi_min);
+      // find nearest k matrices
+      index1 = floor((NO_OF_EQUILIBRIUM_POSITIONS-1)*phi_normalised);
+      index2 = ceil((NO_OF_EQUILIBRIUM_POSITIONS-1)*phi_normalised);
+    }
+
+    // actual interpolation
+    controller.SetK(k_matrix[index1] + (k_matrix[index2]-k_matrix[index1]) * (9*phi_normalised-index1));
+
+  }
+  double StabControllerMultiLqrNode::LipImuOffset(const double phi){
+    // method interpolates linearly betwen nearest K-matrices
+    int index1;
+    int index2;
+    double phi_normalised;
+
+    if (phi<phi_min) {
+      index1 = 0;
+      index2 = 0;
+      ROS_WARN("LIP: phi out of bounds, using phi min");
+
+    } else if (phi>phi_max) {
+      index1 = NO_OF_EQUILIBRIUM_POSITIONS-1;
+      index2 = NO_OF_EQUILIBRIUM_POSITIONS-1;
+      ROS_WARN("LIP: phi out of bounds, using phi max");
+
+    } else {
+      // map theta to [0, 1]:
+      phi_normalised = (phi-phi_min)/(phi_max-phi_min);
+      // find nearest k matrices
+      index1 = floor((NO_OF_EQUILIBRIUM_POSITIONS-1)*phi_normalised);
+      index2 = ceil((NO_OF_EQUILIBRIUM_POSITIONS-1)*phi_normalised);
+    }
+    // actual interpolation
+    return imu_offset[index1] + (imu_offset[index2]-imu_offset[index1])*((NO_OF_EQUILIBRIUM_POSITIONS-1)*phi_normalised-index1);
+  }
+
+  // Matrix read in methods
+  void StabControllerMultiLqrNode::ReadInMatrixArray(std::string base_filename){
+    // read in k matrices in to k_matrix array
+    for (int i = 0; i < NO_OF_EQUILIBRIUM_POSITIONS; i++) {
+      k_matrix[i] = ReadMatFromCSV(base_filename+std::to_string(i)+".csv", NO_OF_INPUTS, (NO_OF_STATES+NO_OF_DELAYS));
```

Comment:
Ros has some bulit in functionality that could help you here.
For example in the image_undistort repo I also need to load matricies from file and go about it like so
1) the data is stored as a yaml file for example https://github.com/ethz-asl/image_undistort/blob/master/cfg/example_output_config.yaml
2) it is loaded into the ros parameter server by the launchfile https://github.com/ethz-asl/image_undistort/blob/master/launch/basic_undistort_example.launch#L11
3) the code loads the rosparam to an xmlrpc https://github.com/ethz-asl/image_undistort/blob/master/src/camera_parameters.cpp#L12
4) it is converted to an eigen matrix https://github.com/ethz-asl/image_undistort/blob/master/include/image_undistort/camera_parameters.h#L46

While a little complex, it means you don't have to worry about file paths and is fairly flexible. It also comes with all of the parsing and error checking built into the yaml loader in ros which makes debugging a mistake in the loaded file easier